# French Red wine price predicting
I will use datase of French Red wine, scrapped from vivino
NOTE : DONT RUN THIS, the unclean data france_wines_type1.csv where not pushed (only a subset of the clean data used in the next part are in).


In [17]:

import pandas as pd
import numpy as np
import os
import sys
from pathlib import Path

## LOT 1.2 Data Loading and first observations

In [18]:
data_path = '../data/france_wines_type1.csv'

if not os.path.exists(data_path):
    print(f"❌ Error: Data file not found at {data_path}")
    sys.exit(1)

df = pd.read_csv(data_path)

print(f"  ✅ Loaded: {data_path}")
print(f"  📈 Shape: {df.shape[0]:,} rows × {df.shape[1]} columns\n")
df

  ✅ Loaded: ../data/france_wines_type1.csv
  📈 Shape: 91,060 rows × 21 columns



,id,id_vintage,wine,type_id,winery,url,global_rating,global_rating_count,vintage_year,vintage_rating,...,price_availability,structure_acidity,structure_fizziness,structure_intensity,structure_sweetness,structure_tannin,region,country,grapes,food_pairings
0,5776748,161549629,La Caumette,1,Domaine Lafage,https://www.vivino.com/FR/en/la-caumette/w/577...,4.3,875,2019,4.3,...,24.90,4.0,NaN,4.0,1.0,4.0,languedoc-roussillon,france,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2...","[(4, 'beef'), (5, None), (9, 'veal')]"
1,1145410,1417343,Saint-Émilion Grand Cru,1,Clos des Menuts,https://www.vivino.com/FR/en/saint-emilion-gra...,4.0,7155,2002,4.3,...,30.00,3.5,NaN,4.5,1.0,4.0,st-emilion-grand-cru,france,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran...","[(4, 'beef'), (8, 'lamb'), (11, 'game'), (20, ..."
2,1903355,146343800,Terre de Pierres,1,Château Moyau,https://www.vivino.com/FR/en/terre-de-pierres/...,4.1,492,2011,4.3,...,16.90,4.0,NaN,4.0,1.0,4.0,languedoc-la-clape,france,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2...","[(4, 'beef'), (5, None), (9, 'veal')]"
3,1657702,7978335,Saint-Émilion Grand Cru (Grand Cru Classé),1,Château Balestard La Tonnelle,https://www.vivino.com/FR/en/saint-emilion-gra...,4.0,3825,1986,4.3,...,30.00,3.5,NaN,4.5,1.0,4.0,st-emilion-grand-cru,france,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran...","[(4, 'beef'), (8, 'lamb'), (11, 'game'), (20, ..."
4,2353632,167038803,Les Hautes Granges Roaix,1,Domaine des Escaravailles,https://www.vivino.com/FR/en/les-hautes-grange...,4.1,584,2020,4.3,...,24.90,3.5,NaN,4.0,1.0,3.5,cotes-du-rhone-villages-roaix,france,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2...","[(4, 'beef'), (8, 'lamb'), (11, 'game')]"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
91055,1168411,2461672,Clos de Vougeot Grand Cru,1,Méo-Camuzet,https://www.vivino.com/FR/en/clos-de-vougeot-g...,4.4,2891,1999,4.5,...,556.75,4.5,NaN,3.0,1.0,2.5,clos-de-vougeot,france,"[(14, 'pinot-noir')]","[(4, 'beef'), (9, 'veal'), (11, 'game'), (20, ..."
91056,1543831,7168498,Bonnes-Mares Grand Cru,1,Bouchard Père & Fils,https://www.vivino.com/FR/en/bonnes-mares-gran...,4.4,319,2000,4.4,...,477.72,4.5,NaN,3.0,1.0,2.5,bonnes-mares,france,"[(14, 'pinot-noir')]","[(4, 'beef'), (9, 'veal'), (11, 'game'), (20, ..."
91057,1543831,5697894,Bonnes-Mares Grand Cru,1,Bouchard Père & Fils,https://www.vivino.com/FR/en/bonnes-mares-gran...,4.4,319,2010,4.4,...,552.14,4.5,NaN,3.0,1.0,2.5,bonnes-mares,france,"[(14, 'pinot-noir')]","[(4, 'beef'), (9, 'veal'), (11, 'game'), (20, ..."
91058,1180656,2077633,Chambertin Grand Cru,1,Bouchard Père & Fils,https://www.vivino.com/FR/en/chambertin-grand-...,4.3,334,2002,4.3,...,564.14,4.5,NaN,3.0,1.0,2.5,chambertin,france,"[(14, 'pinot-noir')]","[(4, 'beef'), (9, 'veal'), (11, 'game'), (20, ..."


In [19]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91060 entries, 0 to 91059
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    91060 non-null  int64  
 1   id_vintage            91060 non-null  int64  
 2   wine                  91060 non-null  object 
 3   type_id               91060 non-null  int64  
 4   winery                91040 non-null  object 
 5   url                   91060 non-null  object 
 6   global_rating         91060 non-null  float64
 7   global_rating_count   91060 non-null  int64  
 8   vintage_year          91058 non-null  object 
 9   vintage_rating        91060 non-null  float64
 10  vintage_rating_count  91060 non-null  int64  
 11  price_availability    91060 non-null  float64
 12  structure_acidity     91056 non-null  float64
 13  structure_fizziness   0 non-null      float64
 14  structure_intensity   91056 non-null  float64
 15  structure_sweetness

In [20]:
df.describe()


,id,id_vintage,type_id,global_rating,global_rating_count,vintage_rating,vintage_rating_count,price_availability,structure_acidity,structure_fizziness,structure_intensity,structure_sweetness,structure_tannin
count,9.106000e+04,9.106000e+04,91060.0,91060.000000,91060.000000,91060.00000,91060.000000,91060.000000,91056.000000,0.0,91056.000000,91056.000000,91056.000000
mean,2.331605e+06,1.246171e+08,1.0,4.000472,3453.495959,4.01072,802.324380,125.986056,4.006254,NaN,3.547767,1.030860,3.321478
std,2.456030e+06,6.991153e+07,0.0,0.266082,8116.990399,0.27111,3230.509348,760.961976,0.500378,NaN,0.935299,0.120324,1.019341
min,9.810000e+02,3.356000e+03,1.0,2.600000,25.000000,2.60000,25.000000,4.000000,2.500000,NaN,1.000000,1.000000,1.000000
25%,1.136028e+06,3.587591e+07,1.0,3.800000,255.000000,3.80000,65.000000,22.967500,3.500000,NaN,3.000000,1.000000,2.500000
50%,1.389505e+06,1.627844e+08,1.0,4.000000,730.000000,4.00000,169.000000,47.330000,4.000000,NaN,4.000000,1.000000,3.500000
75%,2.407248e+06,1.720513e+08,1.0,4.200000,2458.250000,4.20000,463.000000,107.000000,4.500000,NaN,4.500000,1.000000,4.000000
max,1.317291e+07,1.811036e+08,1.0,4.700000,111465.000000,4.80000,111450.000000,171307.000000,5.000000,NaN,5.000000,1.500000,5.000000


In [21]:
# Sort columns: numeric first, then non-numeric
numeric_cols = df.select_dtypes(include=['number']).columns
categorical_cols = df.select_dtypes(exclude=['number']).columns
ordered_cols = list(numeric_cols) + list(categorical_cols)

# Explore unique values, missing values, and data types for each column
for col in ordered_cols:
    dtype = df[col].dtype
    missing = df[col].isna().sum()
    unique = df[col].nunique()
    sample = df[col].dropna().unique()[:5]

    print(f"{col} ({dtype})")
    print(f"  Unique: {unique:,} | Missing: {missing:,}")
    print(f"  Sample: {list(sample)}")
    print()


id (int64)
  Unique: 27,043 | Missing: 0
  Sample: [np.int64(5776748), np.int64(1145410), np.int64(1903355), np.int64(1657702), np.int64(2353632)]

id_vintage (int64)
  Unique: 91,060 | Missing: 0
  Sample: [np.int64(161549629), np.int64(1417343), np.int64(146343800), np.int64(7978335), np.int64(167038803)]

type_id (int64)
  Unique: 1 | Missing: 0
  Sample: [np.int64(1)]

global_rating (float64)
  Unique: 22 | Missing: 0
  Sample: [np.float64(4.3), np.float64(4.0), np.float64(4.1), np.float64(4.2), np.float64(3.9)]

global_rating_count (int64)
  Unique: 5,847 | Missing: 0
  Sample: [np.int64(875), np.int64(7155), np.int64(492), np.int64(3825), np.int64(584)]

vintage_rating (float64)
  Unique: 23 | Missing: 0
  Sample: [np.float64(4.3), np.float64(4.2), np.float64(4.1), np.float64(4.0), np.float64(3.9)]

vintage_rating_count (int64)
  Unique: 4,267 | Missing: 0
  Sample: [np.int64(174), np.int64(107), np.int64(62), np.int64(38), np.int64(36)]

price_availability (float64)
  Unique: 14

In [22]:
# Detect non-numeric entries in vintage_year
invalid_mask = ~df['vintage_year'].astype(str).str.fullmatch(r'\d{4}')
invalid_values = df.loc[invalid_mask, 'vintage_year']

print("🧩 Non-numeric / invalid 'vintage_year' entries:")
print(invalid_values.value_counts())
# print(f"\nTotal invalid: {invalid_values.count()} / {len(df)}")


🧩 Non-numeric / invalid 'vintage_year' entries:
vintage_year
N.V.      29
2022.0    15
2020.0    13
2019.0    12
2021.0    12
2018.0     8
2016.0     6
2023.0     5
1993.0     4
2015.0     4
2011.0     3
2014.0     2
2005.0     1
2001.0     1
2024.0     1
2006.0     1
2000.0     1
2004.0     1
2017.0     1
2012.0     1
Name: count, dtype: int64


In this dataset we have both numerical and categorical :
### **Numerical Columns**
* **id**: Unique wine identifier across all vintages
* **id_vintage**: Unique identifier for specific wine-vintage combination
* **type_id**: Wine type category (1 = red wine)

* **global_rating**: Average rating across all vintages (scale 1-5)
* **global_rating_count**: Total number of user ratings for the wine
* **vintage_rating**: Average rating for specific vintage (scale 1-5)
* **vintage_rating_count**: Number of user ratings for specific vintage

* **price_availability**: Wine price in euros (€)

* **structure_acidity**: Acidity level (scale 1-5, where 5 = most acidic) (soft-acid)
* **structure_fizziness**: Carbonation level (scale 1-5, all null for red wines)
* **structure_intensity**: Flavor intensity (scale 1-5, where 5 = most intense) (light-intense)
* **structure_sweetness**: Sweetness level (scale 1-5, where 5 = sweetest) (dry-sweet)
* **structure_tannin**: Tannin level (scale 1-5, where 5 = highest tannin)

## **Categorical Columns**
* **wine**: Commercial name of the wine
* **winery**: Producer/winery name
* **url**: Vivino product page URL
* **region**: French wine-producing region (e.g., languedoc-roussillon, saint-estephe)
* **country**: Country of origin (france)
* **vintage_year**: Year of harvest/production

* **grapes**: List of grape varieties with IDs in format [(id, 'name'), ...]
* **food_pairings**: Recommended food pairings with IDs in format [(id, 'name'), ...]

### Handling missing values
There are few values missing, so let's just remove those out of the dataset.

**Negligible Missing Data (can safely ignore):**
- **winery**: 20 missing (0.02%)
- **vintage_year**: 2 missing (0.002%)
- **structure_acidity, structure_intensity, structure_sweetness, structure_tannin**: 4 missing each (0.004%)

**Moderate Missing Data:**
- **grapes**: 4,322 missing (4.7%) - grape composition not documented for some wines

**Complete Missing Data:**
- **structure_fizziness**: 91,060 missing (100%) - expected, as red wines aren't sparkling; **can be dropped**

**Summary:**
Apart from `structure_fizziness` (which should be dropped), all other missing values are minimal (<5%). The dataset is quite complete overall.
I'll replace grape missing values with "Unknown" for now.

I think few columns are not particularly interesting like:
  - food_pairings
  - url
  - country: all from France
  - id and id_vintage : id from vivino, does not make sens to keep them here
  - global_rating and global_rating_count, i'm interest in each wine bottle per year
So i'll drop them
Few categories are identifiers like id, id_vintage, wine and winery.
 vintage_year is an object instead of a int or float



In [29]:
# Drop structure_fizziness column (100% missing)
df_france = df.drop(columns=['structure_fizziness'])
df_france = df_france.drop(columns=['food_pairings'])
df_france = df_france.drop(columns=['url'])
df_france = df_france.drop(columns=['country'])
df_france = df_france.drop(columns=['id'])
df_france = df_france.drop(columns=['type_id'])
df_france = df_france.drop(columns=['id_vintage'])
df_france = df_france.drop(columns=['global_rating'])
df_france = df_france.drop(columns=['global_rating_count'])
df['vintage_year'] = df['vintage_year'].replace('N.V.', np.nan) # replace year missing with nan
print(f"✅ Dropped 'structure_fizziness','food_pairings','url', 'country','country','id', 'id_vintage','global_rating_count','global_rating'")

# Remove rows with any missing values
df_france['grapes'] = df_france['grapes'].fillna('Unknown')
df_france_clean = df_france.dropna()
print(f"✅ Removed rows with missing values")

print(f"📊 Original shape: {df_france.shape[0]:,} rows × {df_france.shape[1]} columns")
print(f"📊 Clean shape: {df_france_clean.shape[0]:,} rows × {df_france_clean.shape[1]} columns")
print(f"🗑️  Removed: {df_france.shape[0] - df_france_clean.shape[0]:,} rows ({((df_france.shape[0] - df_france_clean.shape[0]) / df_france.shape[0] * 100):.1f}%)")
print(f"\n✓ No missing values remaining: {df_france_clean.isna().sum().sum() == 0}")

# Save the clean dataset
output_path = '../data/france_wines_type1_clean.csv'
df_france_clean.to_csv(output_path, index=False)
print(f"\n💾 Saved clean dataset to: {output_path}")

✅ Dropped 'structure_fizziness','food_pairings','url', 'country','country','id', 'id_vintage','global_rating_count','global_rating'
✅ Removed rows with missing values
📊 Original shape: 91,060 rows × 12 columns
📊 Clean shape: 91,005 rows × 12 columns
🗑️  Removed: 55 rows (0.1%)

✓ No missing values remaining: True

💾 Saved clean dataset to: ../data/france_wines_type1_clean.csv


In [30]:
df_france_clean

,wine,winery,vintage_year,vintage_rating,vintage_rating_count,price_availability,structure_acidity,structure_intensity,structure_sweetness,structure_tannin,region,grapes
0,La Caumette,Domaine Lafage,2019,4.3,174,24.90,4.0,4.0,1.0,4.0,languedoc-roussillon,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2..."
1,Saint-Émilion Grand Cru,Clos des Menuts,2002,4.3,107,30.00,3.5,4.5,1.0,4.0,st-emilion-grand-cru,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran..."
2,Terre de Pierres,Château Moyau,2011,4.3,62,16.90,4.0,4.0,1.0,4.0,languedoc-la-clape,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2..."
3,Saint-Émilion Grand Cru (Grand Cru Classé),Château Balestard La Tonnelle,1986,4.3,38,30.00,3.5,4.5,1.0,4.0,st-emilion-grand-cru,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran..."
4,Les Hautes Granges Roaix,Domaine des Escaravailles,2020,4.3,36,24.90,3.5,4.0,1.0,3.5,cotes-du-rhone-villages-roaix,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2..."
...,...,...,...,...,...,...,...,...,...,...,...,...
91055,Clos de Vougeot Grand Cru,Méo-Camuzet,1999,4.5,108,556.75,4.5,3.0,1.0,2.5,clos-de-vougeot,"[(14, 'pinot-noir')]"
91056,Bonnes-Mares Grand Cru,Bouchard Père & Fils,2000,4.4,319,477.72,4.5,3.0,1.0,2.5,bonnes-mares,"[(14, 'pinot-noir')]"
91057,Bonnes-Mares Grand Cru,Bouchard Père & Fils,2010,4.4,319,552.14,4.5,3.0,1.0,2.5,bonnes-mares,"[(14, 'pinot-noir')]"
91058,Chambertin Grand Cru,Bouchard Père & Fils,2002,4.3,334,564.14,4.5,3.0,1.0,2.5,chambertin,"[(14, 'pinot-noir')]"


In [25]:
df_france_clean.describe()

,type_id,vintage_rating,vintage_rating_count,price_availability,structure_acidity,structure_intensity,structure_sweetness,structure_tannin
count,91034.0,91034.000000,91034.000000,91034.000000,91034.000000,91034.000000,91034.000000,91034.000000
mean,1.0,4.010812,802.160511,126.009975,4.006300,3.547718,1.030868,3.321484
std,0.0,0.271021,3229.501566,761.069012,0.500389,0.935299,0.120338,1.019347
min,1.0,2.600000,25.000000,4.000000,2.500000,1.000000,1.000000,1.000000
25%,1.0,3.800000,65.000000,22.970000,3.500000,3.000000,1.000000,2.500000
50%,1.0,4.000000,169.000000,47.330000,4.000000,4.000000,1.000000,3.500000
75%,1.0,4.200000,463.000000,107.020000,4.500000,4.500000,1.000000,4.000000
max,1.0,4.800000,111450.000000,171307.000000,5.000000,5.000000,1.500000,5.000000


In [31]:
# Extract 1000 random samples from the clean dataset
df_sample = df_france_clean.sample(n=1000, random_state=42)

print(f"📊 Sampled dataset: {df_sample.shape[0]:,} rows × {df_sample.shape[1]} columns")

# Save sample to CSV
sample_path = '../data/france_wines_sample_1000.csv'
df_sample.to_csv(sample_path, index=False)

print(f"💾 Saved sample to: {sample_path}")

# Check file size
import os
file_size_mb = os.path.getsize(sample_path) / (1024 * 1024)
print(f"📁 File size: {file_size_mb:.2f} MB")

df_sample.head()

📊 Sampled dataset: 1,000 rows × 12 columns
💾 Saved sample to: ../data/france_wines_sample_1000.csv
📁 File size: 0.15 MB


,wine,winery,vintage_year,vintage_rating,vintage_rating_count,price_availability,structure_acidity,structure_intensity,structure_sweetness,structure_tannin,region,grapes
16381,Margaux,Château Rauzan-Gassies,1993,3.9,33,72.00,4.5,4.0,1.0,4.5,margaux,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran..."
21156,Saint-Estephe,Cos d'Estournel,2005,4.2,171,404.00,4.5,4.5,1.0,5.0,saint-estephe,"[(2, 'cabernet-sauvignon'), (3, 'cabernet-fran..."
3313,Authentique Côtes du Roussillon Rouge,Domaine Lafage,2023,3.9,70,8.85,4.0,4.0,1.0,4.0,cotes-du-roussillon,"[(1, 'shiraz-syrah'), (8, 'grenache-noir'), (2..."
81380,Mercurey 1er Cru 'Les Naugues',Domaine François Raquillet,2019,4.0,25,36.00,4.0,2.0,1.0,2.0,mercurey-les-naugues,"[(14, 'pinot-noir')]"
60038,Châteauneuf-du-Pape La Cuvée du Papet,Clos du Mont-Olivet,2016,4.2,138,69.50,3.0,5.0,1.0,3.0,chateauneuf-du-pape,"[(1, 'shiraz-syrah'), (22, 'mourvedre'), (32, ..."


In [32]:
# Sort columns: numeric first, then non-numeric
numeric_cols = df_france_clean.select_dtypes(include=['number']).columns
categorical_cols = df_france_clean.select_dtypes(exclude=['number']).columns
ordered_cols = list(numeric_cols) + list(categorical_cols)

# Explore unique values, missing values, and data types for each column
for col in ordered_cols:
    dtype = df_france_clean[col].dtype
    missing = df_france_clean[col].isna().sum()
    unique = df_france_clean[col].nunique()
    sample = df_france_clean[col].dropna().unique()[:5]

    print(f"{col} ({dtype})")
    print(f"  Unique: {unique:,} | Missing: {missing:,}")
    print(f"  Sample: {list(sample)}")
    print()


vintage_rating (float64)
  Unique: 23 | Missing: 0
  Sample: [np.float64(4.3), np.float64(4.2), np.float64(4.1), np.float64(4.0), np.float64(3.9)]

vintage_rating_count (int64)
  Unique: 4,263 | Missing: 0
  Sample: [np.int64(174), np.int64(107), np.int64(62), np.int64(38), np.int64(36)]

price_availability (float64)
  Unique: 14,997 | Missing: 0
  Sample: [np.float64(24.9), np.float64(30.0), np.float64(16.9), np.float64(30.5), np.float64(27.9)]

structure_acidity (float64)
  Unique: 6 | Missing: 0
  Sample: [np.float64(4.0), np.float64(3.5), np.float64(4.5), np.float64(3.0), np.float64(2.5)]

structure_intensity (float64)
  Unique: 8 | Missing: 0
  Sample: [np.float64(4.0), np.float64(4.5), np.float64(5.0), np.float64(3.5), np.float64(2.0)]

structure_sweetness (float64)
  Unique: 2 | Missing: 0
  Sample: [np.float64(1.0), np.float64(1.5)]

structure_tannin (float64)
  Unique: 9 | Missing: 0
  Sample: [np.float64(4.0), np.float64(3.5), np.float64(5.0), np.float64(3.0), np.float64(4.5)